build_tools/cmake/build_docs.sh

```diff
@@ -58,3 +59,61 @@ scripts/update_op_coverage.py ${BUILD_DIR}
 # Update e2e_coverage.md
 PYTHON_BIN=`which python3` scripts/update_e2e_coverage.py ${BUILD_DIR}
 
+# Generate IR conversion flow examples
+
+# The list of input IR examples
+declare -a ir_examples=(
+  iree/samples/ops/mhlo-dot.mlir
+  iree/samples/ops/mhlo-conv.mlir
+)
+# The list of compiler targets
+declare -a compiler_targets=(
+  llvm-ir
+  vmla
+  vulkan-spirv
+)
+
+mkdir -p ${BUILD_DIR}/doc/ir_examples
+for example in "${ir_examples[@]}"; do
+  for target in "${compiler_targets[@]}"; do
+    example_name=$(basename ${example%.mlir})
+    filepath=${BUILD_DIR}/doc/ir_examples/${example_name}-${target}.md
+    title="${example_name} IR conversion example for ${target} backend"
+
+    echo -e "# ${title}\n" > ${filepath}
+
+    preamble="
+This is an IR conversion example auto-generated by running \`iree-opt\`
+over [${example_name}](https://github.com/google/iree/tree/main/${example}).
+
+### The command
+"
+    echo -e "${preamble}" >> ${filepath}
+    echo -e -n '```shell' >> ${filepath}
+
+    cmd="
+iree-opt -iree-transformation-pipeline \\
+  -iree-hal-target-backends=${target} \\
+  --print-ir-after-all \\
+  -mlir-disable-threading \\
+  ${example}
+"
+    echo -e "${cmd}" >> ${filepath}
+    echo -e '```' >> ${filepath}
+
+    echo -e "### Input IR\n" >> ${filepath}
+    echo -e '```' >> ${filepath}
+    cat ${example} >> ${filepath}
+
+    tmpfile=$(mktemp)
+    ${BUILD_DIR}/iree/tools/iree-opt \
+      -iree-transformation-pipeline \
+      -iree-hal-target-backends=${target} \
+      -print-ir-after-all \
```

Comment:
`print-ir-after-all` is certainly easiest to set up, but I have a few concerns with using it for demonstration:

* The full output is quite lengthy. My browser scroll bar is so tiny!
* On https://antiagainst.github.io/iree/ir-examples/mhlo-conv-llvm-ir, `Canonicalizer` runs 21 times, `CSE` runs 16 times. Sometimes in a loop: https://antiagainst.github.io/iree/ir-examples/mhlo-conv-llvm-ir#ir-dump-after-canonicalizer-13. That's a lot of noise, even if technically complete.
* Some of these passes and phases would really benefit from descriptions. We could pull the descriptions from command line help text for each pass, but hand-written comments with the full context of running a pass pipeline for a specific model could be more helpful.

Reply:
I'd prefer to keep the full picture of all the passes run there so it's easier to contrast with the pipeline in source and also clear to see how the IR is being transformed. Deleting certain stages might be confusing for following the steps there.

Good point regarding adding comments to the passes! Also probably linking to the source code.. It needs some shell script magic though I think we can defer to the next step. :)